pandaserver/dataservice/closer.py

```diff
@@ -0,0 +1,400 @@
+"""
+update dataset DB, and then close dataset and start Activator if needed
+
+"""
+
+import re
+import sys
+from typing import Dict, List
+
+from pandacommon.pandalogger.PandaLogger import PandaLogger
+
+from pandaserver.config import panda_config
+from pandaserver.dataservice import Notifier
+from pandaserver.dataservice.Activator import Activator
+from pandaserver.taskbuffer import EventServiceUtils
+
+# logger
+_logger = PandaLogger().getLogger("closer")
+
+
+def init_logger(p_logger: PandaLogger) -> None:
+    """
+    Redirect logging to parent as it doesn't work in nested threads
+
+    Args:
+        p_logger (PandaLogger): The parent logger.
+    """
+    # redirect logging to parent as it doesn't work in nested threads
+    global _logger
+    _logger = p_logger
+    Notifier.initLogger(_logger)
+
+
+class Closer:
+    """
+    Update dataset DB, and then close dataset and start Activator if needed
+    """
+
+    # constructor
+    def __init__(self, taskBuffer, destination_db_blocks: List[str], job, panda_ddm: bool = False,
+                 dataset_map: Dict = {}) -> None:
+        """
+        Constructor
+
+        Args:
+            task_buffer: Task buffer.
+            destination_db_blocks (List[str]): Destination DB blocks.
+            job: Job.
+            panda_ddm (bool, optional): Panda DDM flag. Defaults to False.
+            dataset_map (Dict, optional): Dataset map. Defaults to {}.
+        """
+        self.task_buffer = taskBuffer
+        self.destination_db_blocks = destination_db_blocks
+        self.job = job
+        self.panda_id = job.PandaID
+        self.panda_ddm = panda_ddm
+        self.site_mapper = None
+        self.dataset_map = dataset_map
+        self.all_sub_finished = None
+
+    def start(self) -> None:
+        """
+        To keep backward compatibility
+        """
+        self.run()
+
+    def join(self) -> None:
+        """
+        Join
+        """
+        pass
+
+    # main
+    def run(self):
+        """
+        Main
+        """
+        try:
+            _logger.debug(f"{self.panda_id} Start {self.job.jobStatus}")
+            flag_complete = True
+            top_user_ds_list = []
+            using_merger = False
+            disable_notifier = False
+            first_indv_ds = True
+            final_status_ds = []
+            for destination_db_block in self.destination_db_blocks:
+                ds_list = []
+                _logger.debug(f"{self.panda_id} start {destination_db_block}")
+                # ignore tid datasets
+                if re.search("_tid[\d_]+$", destination_db_block):
```

Comment:
There is too much dataset naming convention (tid, hc_test, user_gangarbt, _sub, .lib...) throughout the code and it makes it difficult to follow. Make a summary of the dataset naming conventions, add some comments, extract to short functions,...?